File: 242-valid-anagram/valid-anagram.swift
class Solution {
    func isAnagram(_ s: String, _ t: String) -> Bool {
        
        // 1. Put all characters into Hash [Character: NumberOfCharacters]
        // 2. Loop through all characters in t
        // 3. Check that the character exists in our hash
        // 4. if not exists return false
        // 5. If it exists subtract 1 from corresponding value
        // 6. Repeat loop

        var allowedCharacters = s.reduce(into: [Character: Int]()) { acc, character in
            acc[character] = (acc[character] ?? 0) + 1
        }

        for character in t {

            guard 
                let numberOfCharacters = allowedCharacters[character],
                numberOfCharacters > 0
            else {
                return false
            }

            let newNumberOfCharacters = numberOfCharacters - 1

            if newNumberOfCharacters <= 0 {
                allowedCharacters.removeValue(forKey: character)
                continue
            }

            allowedCharacters[character] = newNumberOfCharacters
        }

        return allowedCharacters.isEmpty
    }
}

File: 242-valid-anagram/valid-anagram.py
import functools

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        available_characters = {}

        for character in s:
            available_characters[character] = available_characters.get(character, 0) + 1

        for character in t:
            number_of_characters = available_characters.get(character, 0)

            if number_of_characters == 0:
                return False

            number_of_characters -= 1
            available_characters[character] = number_of_characters

            if number_of_characters == 0:
                del available_characters[character]

        return len(available_characters) == 0